### PR TITLE
Create filters

### PR DESCRIPTION
**Pull request by zrpnr**
While working on the filters I realized the defaults for a parsed url were incomplete, I separated those methods out of the `Location` since they don't depend on state or context.

Added tests and new methods for `include` and `fields`.

This should also set some foundations for the "history" feature.

### CODE REVIEW

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: Maybe this could always be a `Set`?

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: When would this be false?

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: When would the indexOf condition be false?

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: I like these two lines. It would be tempting to use a regex, but I think the more mechanical index and substring based approach is appropriate because this syntax is defined by the spec.

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js:33]: `sort` has the same syntax as include. Maybe we can `s/parseInclude/parseListParameter` and `s/parseFields/parseDictionaryParameter` and after that, `parseDictionaryParameter` could reuse `parseListParameter` for each instance of a `field[*]` query parameter, since that is a list too.

File: src/lib/url.test.js
Comment (gabesullice) [inline src/lib/url.test.js:162]: These are fantastic! I need to get better about writing this tests too. I've let some bad habits grow after so long w/o writing JS.

**Review by gabesullice — COMMENTED** [review]
This PR looks great! Why is it still a draft? I think we could merge it as-is.

I asked a few questions for my understanding (not to suggest changes).

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js]: I was thinking when there is user input- like a url pasted into the browser bar we'd want to validate that it was a correct filter. This probably doesn't go far enough to really do that. I was mostly adding it since I referred to the indexOf directly below.

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js]: Do you think we will need a better way to "process" out these keys- specifically with the more nested syntax of filters `filter[author-filter][condition][path]=uid.id` -here it works but may need to be more robust.

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js:33]: I like that a lot, using a more generic name. I hadn't gotten to sort yet but it seemed like it had some more complexity, such as the "normal" syntax and with sort order.

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js]: Ah never I guess- it is always an instance of `URLSearchParams` - I'll remove this.

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js]: Do you think there's an advantage? It will never change, no need to check for unique properties.

**Comment by zrpnr** [comment]
I left this a draft because I'd intended to work on "filter" but I worked on the existing parts first. Filter seemed to be quite a bit more complex. We could certainly merge this much and then tackle filters (and sort) next.

**Comment by zrpnr** [comment]
Made the changes you suggested, except for making the array of query types a `Set` since its only purpose here is to be filter and mapped over- which are Array methods.

File: src/lib/url.js
Comment (gabesullice) [reply src/lib/url.js]: What do you mean by a better way? We'll definitely need to modify this or write something else to handle the deeper "nesting". There might be an npm package, but it also wouldn't be terribly difficult to write ourselves. Could be fun to pair on actually.

File: src/lib/url.js
Comment (gabesullice) [reply src/lib/url.js:33]: Ugh, good call. I totally forgot about the "normal" thing on https://www.drupal.org/docs/8/modules/jsonapi/sorting.

I don't know why we support that non-spec'd way when the spec seems perfectly sufficient. It's probably my fault and I've just forgotten that it is lol; maybe a leftover from https://www.drupal.org/project/entityqueryapi

File: src/lib/url.js
Comment (gabesullice) [reply src/lib/url.js]: I thought that we were converting it to set in multiple places but I only found `utils.js::checkIncludesPath`. It'd make that a teeny bit simpler. Not worth blocking this merge on.

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: `s/include/value` now that the method name isn't include specific.

File: src/lib/url.js
Comment (gabesullice) [inline src/lib/url.js]: I think `'fields'` should become an argument to this function.

**Review by gabesullice — CHANGES_REQUESTED** [review]
Love where this went. I had a few nits. I'll pull this branch and push some commits to it since I don't think you're online yet.

File: src/lib/url.js
Comment (zrpnr) [reply src/lib/url.js]: I had just pulled `queryParams` out of the `compileJsonApiUrl` method, it had just been a hardcoded array of those values before.
Its only use is 
```const queryString = queryParams.filter( ... ).map( ... ).join( ... )```

**Review by gabesullice — APPROVED** [review]
@zrpnr, I'll let you merge this since I pushed two commits to it. I think this is ready to go.

**Comment by zrpnr** [comment]
I'm glad you renamed those and variable names, that will make more sense when we reuse them for sort and filters. Merging, thanks!

**Pull request merged**
